LochNessBuilder/Extensions/ExpressionExtensions.cs

```diff
@@ -7,8 +7,12 @@ public static class ExpressionExtensions
     {
         public static string GetMemberName<TInstance, TProp>(this Expression<Func<TInstance, TProp>> selector)
         {
-            var selectorBody = (MemberExpression)selector.Body;
-            return selectorBody.Member.Name;
+            var unaryExpression = selector.Body as UnaryExpression;
```

Comment:
In a similar vein, a quick comment explaining the use cases here might be helpful.

Reply:
Thoughts on this as a Doc block?

```
        /* ========================
         * ==  Unary Conversions ==
         * ========================
         *
         * Most calls looks something like this:
         *     `.With(obj => obj.IntProp, 2)`
         *
         * But some calls might look like this:
         *     `.With(obj => obj.ShortProp, 2019)`
         * where the prop type and the argument are not the same, but have implicit casts available.
         *
         * Unfortunately, then compiler turns this into:
         *     `.With<int>(obj => (int)(obj.ShortProp), 2019)`
         * i.e. rather than cast the argument, it assumes an implicit cast on the selector expression.
         * This is the scenario that we're trying to handle with these two methods.
         * One to extract the `obj.ShortProp` out in either case, and one to explicitly detect the casting case
         * as we need to know that the `TProp` that we think we have isn't actually correct.
         * (i.e. `obj.ShortProp` is a `short`, not the `int` that we have in `TProp`.)
         *
         * In most *realistic* situations, we will be fine to convert the value, rather than the Property.
         * But it is entirely possible to write code that compiles, but which can't execute:
         *     `.With(obj => obj.ChildTypeProp, new ParentType())`
         * As in the obj.ShortProp example, this will compile, as it can be interpretted as:
         *     `.With<ParentType>(obj => (ParentType)(obj.ChildTypeProp), new ParentType())`
         * but it cannot possibly succeed, as it is not possible to assign a `new ParentType()` to `obj.ChildTypeProp`.
         * The relevant ArgumentException will be thrown, on the `.With()`, and the user should be able to fix the mis-usages.
         */
```

Reply:
Really great doc - thanks Mike :D